### DISCUSSION TITLE
Release process

### DISCUSSION THREAD

**Discussion by Nexucis** [perses/perses#331] [Design Discussions]
I’m opening this discussion so we can talk about the release process we should put in place.

I have thought about different scenarios, there are certainly others don’t worry !

Currently we will have to release: 

* The perses server:
    * Binaries
        * `Perses` (which contains backend and frontend)
        * `Percli` (which is the CLI and delivered in the same archive than the binary perses)
    * A Docker image that contains the binaries
* The libraries contained in the folder `ui/`

@LukeTillman already published the `@perses-dev/*` libraries manually (which is ok!), I’m hoping that at the end of this documentation, he won’t have to do it anymore!

## Proposition 1 - Release everything under the same version

The react libraries and the Perses server will be delivered at the same time, under the same version.

### Pros

* No need to ensure if a lib is compatible with the server.
* A single tag for a single release

### Cons

* Changelog could be huge and not simple to read.
* It won’t be fast to release the libraries 
* When a lib or the server has a critical bug, we would have to release everything just to fix this part.

## Proposition 2 - Release each part with a different tag / different version

The idea here is to release a single lib with a dedicated tag which can follow this pattern: `<name of the lib>-<version>`. 
For example if we want to deliver the lib `@perses-dev/core`, we would tag the repo with `perses-dev-core-0.2.0`

The exception of this rule would be when we want to deliver the perses server. We will tag the repo with this pattern `v<version>`. It’s a constraint coming with the Go mod versioning that enforces the usage of the semantic versioning. And the repo is also providing some Go libs under the folder `pkg`. So to be consumed, we would need to have a tag digested by Go.

### Pros

* Changelog would be splited and so easy to read. But there would be many different Changelog. It can be painful to manage.
* We are able to release fast the different libs and server

### Cons

* A lot of different tags.
* When we will create a new lib, we will have to modify the release process to include this lib.
* Compatibility between the libs and with the server will be complicated to manage.

##  Proposition 3 - All libs are delivered under a same tag

Server is still delivered under the tag vX.X.X.
But all libs are delivered under a single different tag. Like perses-lib-<version>.

### Pros

* We would just have two different Changelog. One for the server, one for all libs.
* Just two tag pattern
* We can still release faster the frontend libraries and it doesn’t depend on the server release.
* We can ensure that all libs are compatible between each others under the same version.

### Cons

* For a critical bug fix in a particular lib, we will release all lib at once. But maybe it’s not a big issue if it doesn’t occur often.
* Compatibility between the server and the lib is not necessary guaranteed

## Notes 

* Regarding for the proposition 2 and 3:

To guarantee and to understand if the lib is compatible with the server, perhaps we could have this pattern for the lib version : `<V.W>.<X.Y.Z>`. We would have 5 numbers. `V.W` are the two first numbers of the server version. `X.Y.Z` are the release version of the actual lib.

Basically it would mean that the lib under the version `X.Y.Z` is compatible for sure with the version `V.W` of the server.

* Once we are agree on the procedure, we should provide a documentation explaining the release process. Probably we can take some inspiration from Prometheus: https://github.com/prometheus/prometheus/blob/main/RELEASE.md (in terme of pre-step before the actual release like updating the dependencies, the creation of a dedicated branch …etc.)

**Comment by Nexucis** [comment]
Based on the discussion the Core team had, we will put in place the first proposition : release everything under the same release. We realised that the server should be released when a new library is released because everything is packaged through the `react-app` inside the binary.

We will have single Changelog for the moment. In case, the release process doesn't fit our requirement, we will change it.
At some point it will be nice to have a readme on each libraries, so we have a minimum of documentation on NPM.
